File: LevelSDKExampleApp/app/src/main/java/com/theshopatvsp/levelandroidsdk/ble/helper/LeastSquaresExponentialFit.java
package com.theshopatvsp.levelandroidsdk.ble.helper;

/**
 * Created by andrco on 6/13/16.
 */
public class LeastSquaresExponentialFit {
    //returns a, b
    public static double[] fit(double x[], double y[]) {
        double a = 0.0, b = 0.0, part1 = 0.0, part2 = 0.0, part3 = 0.0, part4 = 0.0, part5 = 0.0, part6 = 0.0, part7 = 0.0;

        for (int i = 0; i < x.length; i++) {
            part1 += Math.pow(x[i], 2) * y[i];
            part2 += Math.log(y[i]) * y[i];
            part3 += x[i] * y[i];
            part4 += x[i] * y[i] * Math.log(y[i]);
            part5 += y[i];
            part6 += Math.pow(x[i], 2) * y[i];
        }

        a = ((part1 * part2) - (part3 * part4)) / ((part5 * part6) - Math.pow(part3, 2));
        b = ((part5 * part4) - (part3 * part2)) / ((part5 * part1) - Math.pow(part3, 2));

        return new double[]{Math.exp(a), b};
    }
}

File: LevelIOSSdkExample/LevelIOSSdkExample/LevelIOSSdk/model/helper/LeastSquaresExponentialFit.swift
//
//  LeastSquaresExponentialFit.swift
//  LevelIOSSDK
//
//  Created by Andrew Cook on 7/1/16.
//  Copyright © 2016 TheShop. All rights reserved.
//

import Foundation

class LeastSquaresExponentialFit {
    class func fit(x: [Double], y: [Double]) -> [Double] {
        var a: Double = 0.0, b: Double = 0.0, part1: Double = 0.0, part2: Double = 0.0, part3: Double = 0.0,
            part4: Double = 0.0, part5: Double = 0.0, part6: Double = 0.0
        
      for i in (1...x.count) {
      //for (var i = 0; i < 2; i = i+1) {
            part1 += pow(x[i], 2) * y[i]
            part2 += log(y[i]) * y[i]
            part3 += x[i] * y[i]
            part4 += x[i] * y[i] * log(y[i])
            part5 += y[i]
            part6 += pow(x[i], 2) * y[i]
        }
        
        a = ((part1 * part2) - (part3 * part4)) / ((part5 * part6) - pow(part3, 2))
        b = ((part5 * part4) - (part3 * part2)) / ((part5 * part1) - pow(part3, 2))
        
        return [exp(a), b]
    }
}
